pulp_smash/tests/puppet/api_v2/test_sync_publish.py

```diff
@@ -0,0 +1,650 @@
+# coding=utf-8
+"""Test the API endpoints for puppet `repositories`_.
+
+This module assumes that the tests in
+:mod:`pulp_smash.tests.platform.api_v2.test_repository` hold true. The
+following trees of assumptions are explored in this module::
+
+    It is possible to create an puppet repo with a feed. (CreateTestCase)
+    ├── If a valid feed is given, the sync completes and no errors are
+    │   reported for both valid and ivalid query. (SyncValidFeedTestCase)
+    └── If an invalid feed is given, the sync completes and errors are
+        reported. (SyncInvalidFeedTestCase)
+
+    It is possible to create an puppet repo without a feed. (CreateTestCase)
+    └── It is possible to upload an puppet module to a repository, copy the
+        repository's contents to a second repository, add distributor to both
+        both repositories, publish them. Data integrity of queried modules
+        from pul serverver is also verified (PublishTestCase)
+
+Assertions not explored in this module include:
+
+* Given an puppet repository without a feed, sync requests fail.
+* It is impossible to create two puppet repositories with the same relative URL.  # noqa pylint:disable=line-too-long
+* It is possible to upload a directory of puppet modules to an repository.
+* It is possible to upload content and copy it into multiple repositories.
+* It is possible to get content into a repository via a sync and publish it.
+* Functionality of puppet_install_distributor which install puppet module on pulp server  # noqa pylint:disable=line-too-long
+* Functionality of puppet_file_distributor which makes puppet modules available
+in single directory on pulp server over HTTPS.
+
+.. _repositories:
+   http://pulp.readthedocs.org/en/latest/dev-guide/integration/rest-api/repo/cud.html
+
+"""
+from __future__ import unicode_literals
+
+import requests
+import hashlib
+from itertools import chain
+from pulp_smash.config import get_config
+from pulp_smash.utils import (
+    uuid4,
+    create_repository,
+    delete,
+    handle_response,
+    poll_spawned_tasks,
+)
+from unittest2 import TestCase
+
+_VALID_PUPPET_FEED = 'http://forge.puppetlabs.com'
+_VALID_PUPPET_QUERY = 'pulp-pulp'
+_INVALID_PUPPET_QUERY = 'invalid_puppet_query'
+_INVALID_FEED = 'https://example.com'
+_TASK_END_STATES = ('canceled', 'error', 'finished', 'skipped', 'timed out')
+_REPO_PUBLISH_PATH = '/pulp/puppet/'  # + relative_url + puppet-module
+_CONTENT_UPLOADS_PATH = '/pulp/api/v2/content/uploads/'
+_PUPPET_QUERY_URL = (
+    'http://forge.puppetlabs.com'
+    '/v3/files/pulp-pulp-1.0.0.tar.gz'
+)
+_PUPPET_QUERY_PACKAGE = {
+    'name': 'pulp',
+    'author': 'pulp'
+}
+_PUPPET_DIRECT_URL = 'https://repos.fedorapeople.org/pulp/pulp/demo_repos/puppet_symlink/examplecorp-mymodule-0.1.0.tar.gz'  # noqa pylint:disable=line-too-long
+
+
+def _sync_repo(server_config, href, responses=None):
```

Comment:
I think Issue [#40](https://github.com/PulpQE/pulp-smash/issues/40) discuss exactly this problem. We could move functions that are common for all repository types to `utils.py`. These are so far:
- `_sync_repo()`
- `_get_importers()`
- `_get_units()`
- `_start_content_upload()`
- `_upload_file()`
- `_copy_repo()`
- `_publish_repo()`

We can also move constants `_TASK_END_STATES`, `_CONTENT_UPLOADS_PATH` and list of call report keys [mentioned below](https://github.com/PulpQE/pulp-smash/pull/48#discussion_r46467945) to `constants.py`.